File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2024/Y24D06.kt
package org.gristle.pdxpuzzles.advent.y2024

import kotlinx.coroutines.runBlocking
import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.iteration.parMap
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.toGrid

class Y24D6(input: String) : Day {
    private val lab = input.toGrid()
    private val start = lab.coordOf(lab.indexOf('^'))

    data class State(val pos: Coord, val dir: Nsew, val turned: Boolean)

    private fun move(state: State, obstacle: Coord?): State?  {
        val (pos, dir) = state
        val forward = pos.move(dir)

        // if this is null that means it's gone off the map, which is a valid thing to do!
        return lab.getOrNull(forward)?.let { space ->
            if (space != '#' && forward != obstacle) {
                State(forward, dir, false)
            } else {
                val right = dir.right()
                State(pos, right, true)
            }
        }
    }

    private val goldenPath = generateSequence(State(start, Nsew.NORTH, false)) { move(it, null) }

    override fun part1(): Int = goldenPath.map{ it.pos }.distinct().count()

    override fun part2(): Int = runBlocking {
        val obstacles = mutableSetOf<Coord>()
        goldenPath
            .zipWithNext()
            .toList()
            .filter { (_, next) -> obstacles.add(next.pos) }
            .parMap { (current, next) ->
                val obstacle = next.pos
                val visited = mutableSetOf<State>()
                generateSequence(current) { move(it, obstacle) }
                    .firstOrNull { state ->
                        if (state.turned && state.dir == Nsew.NORTH) {
                            !visited.add(state)
                        } else {
                            false
                        }
                    }
                    ?.let { true }
                    ?: false
            }.count { it }
        }
    }

fun main() = Day.benchmarkDay(Y24D6::class)

//    Class creation: 11ms
//    Part 1: 5444 (13ms)
//    Part 2: 1946 (272ms)
//    Total time: 296ms


@Suppress("unused")
private val test = listOf(
    """....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
""",
)

File: rust/advent/src/bin/2024/y24d06.rs
use std::iter::successors;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use rayon::iter::{ParallelBridge, ParallelIterator};
use utilities::{
    enums::cardinals::Cardinal,
    structs::stopwatch::{ReportDuration, Stopwatch},
};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(24, 6).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
struct State {
    pos: usize,
    dir: Cardinal,
    turned: bool,
}

fn move_dir(pos: usize, dir: Cardinal, width: usize) -> Option<usize> {
    match dir {
        Cardinal::North => pos.checked_sub(width),
        Cardinal::East => Some(pos + 1),
        Cardinal::South => Some(pos + width),
        Cardinal::West => pos.checked_sub(1),
    }
}

fn move_guard(lab: Input, width: usize, state: &State, obstacle: Option<usize>) -> Option<State> {
    let forward = move_dir(state.pos, state.dir, width)?;
    let lab = lab.as_bytes();
    let space = *lab.get(forward)?;
    if space == b'\n' {
        return None;
    }
    if Some(forward) != obstacle && space != b'#' {
        Some(State {
            pos: forward,
            dir: state.dir,
            turned: false,
        })
    } else {
        let right = state.dir.right();
        Some(State {
            pos: state.pos,
            dir: right,
            turned: true,
        })
    }
}

fn golden_path(lab: &str, width: usize, start: State) -> impl Iterator<Item = State> + use<'_> {
    successors(Some(start), move |state| {
        move_guard(lab, width, state, None)
    })
}

fn part1(lab: Input) -> Output {
    let width = lab.find('\n').unwrap() + 1;
    let start = lab.find('^').unwrap();
    let start = State {
        pos: start,
        dir: Cardinal::North,
        turned: false,
    };
    golden_path(lab, width, start)
        .map(|state| state.pos)
        .unique()
        .count()
}

fn part2(lab: &str) -> Output {
    let width = lab.find('\n').unwrap() + 1;
    let start = lab.find('^').unwrap();
    let start = State {
        pos: start,
        dir: Cardinal::North,
        turned: false,
    };
    let mut obstacles = vec![false; lab.len()];
    golden_path(lab, width, start)
        .tuple_windows::<(_, _)>()
        .filter(|(_, next)| {
            let obstacle = next.pos;
            if obstacles[obstacle] {
                false
            } else {
                obstacles[obstacle] = true;
                true
            }
        })
        .par_bridge()
        .filter(|(current, next)| {
            let obstacle = next.pos;
            let mut visited = vec![false; lab.len()];

            successors(Some(*current), move |state| {
                move_guard(lab, width, state, Some(obstacle))
            })
            .any(|state| {
                if state.turned && state.dir == Cardinal::North {
                    if visited[state.pos] {
                        true
                    } else {
                        visited[state.pos] = true;
                        false
                    }
                } else {
                    false
                }
            })
        })
        .count()
}

#[test]
fn default() {
    let input = get_input(24, 6).unwrap();
    assert_eq!(5444, part1(&input));
    assert_eq!(1946, part2(&input));
}

// Input parsed (26μs)
// 1. 5444 (355μs)
// 2. 1946 (4ms)
// Total: 5ms

#[test]
fn examples() {
    let inputs = [r"....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
"];
    assert_eq!(41, part1(inputs[0]));
    assert_eq!(6, part2(inputs[0]));
}
